packages/utilities/web-vitals/src/index.js

```diff
@@ -0,0 +1,130 @@
+const deviceMetrics = {
+  device_mem: null,
+  device_cpu: null,
+  device_effective_connection: null,
+};
+
+const updateWebVitals = ({ name, value }) => {
+  const vitalName = name.toLowerCase();
+  vitals[vitalName] = value;
+};
+
+const updateDeviceMetrics = ({
+  deviceMemory,
+  numberOfLogicalProcessors,
+  effectiveConnectionType,
+}) => {
+  deviceMetrics.device_mem = deviceMemory;
+  deviceMetrics.device_cpu = numberOfLogicalProcessors;
+  deviceMetrics.device_effective_connection = effectiveConnectionType;
+};
+
+const setCurrentUrl = () => {
+  webVitalsBase.url = window.location.href;
+};
+
+const appendReportParams = (reportingEndpoint, reportParams) => {
+  const url = new URL(reportingEndpoint);
+  const reportParamKeys = Object.keys(reportParams);
+  const paramsString = reportParamKeys
+    .map(param => `${param}=${reportParams[param]}`)
+    .join('&');
+
+  return url.search
+    ? `${reportingEndpoint}&${paramsString}`
+    : `${reportingEndpoint}?${paramsString}`;
+};
+
+const sendBeacon = (rawBeacon, reportingEndpoint, reportParams) => {
+  const beacon = JSON.stringify(rawBeacon);
+  const beaconTarget = reportParams
+    ? appendReportParams(reportingEndpoint, reportParams)
+    : reportingEndpoint;
+
+  if (navigator.sendBeacon) {
+    const headers = { type: 'application/reports+json' };
+    const blob = new Blob([beacon], headers);
+    return navigator.sendBeacon(beaconTarget, blob);
+  }
+  return fetch(beaconTarget, {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/reports+json' },
+    body: beacon,
+    mode: 'no-cors',
+  });
+};
+
+const shouldSample = sampleRate => {
+  const randomNumber = Math.floor(Math.random() * 100);
+  return randomNumber <= sampleRate;
+};
+
+const useWebVitals = ({
+  enabled,
+  reportingEndpoint,
+  loggerCallback = noOp,
+  sampleRate = 100,
+  reportParams,
+}) => {
+  let pageLoadTime;
+  const shouldSendVitals = enabled && shouldSample(sampleRate);
+
+  const { effectiveConnectionType } = useNetworkStatus();
+  const { numberOfLogicalProcessors } = useHardwareConcurrency();
+  const { deviceMemory } = useMemoryStatus();
+
+  const sendVitals = async () => {
+    const pageExitTime = Date.now();
+    const pageAge = pageExitTime - pageLoadTime;
+
+    const beacon = [
+      { ...webVitalsBase, age: pageAge, body: { ...vitals, ...deviceMetrics } },
+    ];
+
+    try {
+      await sendBeacon(beacon, reportingEndpoint, reportParams);
+    } catch (error) {
+      loggerCallback(error);
+    }
+  };
+
+  useEvent('pagehide', shouldSendVitals ? sendVitals : noOp);
+
+  useEffect(() => {
+    pageLoadTime = Date.now();
+    setCurrentUrl();
+    updateDeviceMetrics({
+      effectiveConnectionType,
+      numberOfLogicalProcessors,
+      deviceMemory,
+    });
+    getCLS(updateWebVitals);
+    getFID(updateWebVitals);
+    getLCP(updateWebVitals);
+    getFCP(updateWebVitals);
+    getTTFB(updateWebVitals);
```

Comment:
Is it possible to also check for `shouldSendVitals` before even collecting the data?

Reply:
We could do, but I'm worried that may mean we miss the CLS generated by the cookie banner (since `enabled` will be `false` until the cookie banner has been interacted with). Capturing the vitals from the start but not sending them anywhere unless the user accepts the relevant permissions should hopefully mean we can get it to capture that CLS.
This'll likely be something that we'll only be able to play around with and find out/tweak once it's able to be integrated into either webcore or simorgh

Reply:
Oh good point. Depends on how the internal Web APIs work to get the data. They could be collecting the data anyway.